// services/pillarNewspaper.js
var extender    = require("utils/extender");
//var md5         = require("utils/md5");
//var _           = require("lodash");
var spromise       = require("spromise");
var advertisements = require("persistence/advertisements");
var newspapers     = require("persistence/newspapers");
var placements     = require("persistence/placements");
var counters       = require("persistence/counters");

var ready = spromise.when(advertisements.ready);


function pillarNewspaper() {
  this.initialize();
}


function initialize(options) {
  options     = options || {};
  this.active = {};
  this.ready  = ready;
}

function advertisement(data) {
    return spromise(function(resolve){
        var _msg = {
            "id": counters.generateId(),
            "advertisementName": data.advertisementName,
            "advertisementDescription": data.advertisementDescription,
        };
        advertisements.insert(_msg);
        resolve();
    });
}

function advertisementList() {
    return spromise(function(resolve){
    var query = {};

    advertisements.find(query, function(err, advertisements) {
      resolve(advertisements);
    });
  });
}

function newspaper(data) {
    return spromise(function(resolve){
        var _msg = {
            "id": counters.generateId(),
            "newspaperName": data.newspaperName,
        };
        newspapers.insert(_msg);
        resolve();
    });
}

function newspaperList() {
    return spromise(function(resolve){
    var query = {};

    newspapers.find(query, function(err, newspapers) {
      resolve(newspapers);
    });
  });
}

function placeAd(paperID, advertisementID) {
  var queryPaper = {"id":paperID};
  var existingPaper;
  newspapers.findOne(queryPaper,function(err, existingPaper){});
  var queryAdvertisement = {"id":advertisementID};
  var existingAdvertisement;

 advertisements.findOne(queryAdvertisement,function(err, existingAdvertisement){});

    return spromise(function(resolve){
        var _msg = {
            "paperID":paperID,
            "advertisementID":advertisementID,
        };
        placements.insert(_msg);
        resolve();
    });

}

function showPaperAds(paperID) {
    var paper = null;
    var results = [];
    return spromise(function(resolve){
      placements.find({"paperID":paperID},function (err, adverts) {
          resolve(adverts);
      });
    });
}

function showPapersWithAd(adID){
    var ad = null;
    var results = [];
    return spromise(function(resolve){
      placements.find({"advertisementID":adID},function (err, papers) {
          resolve(papers);
      });
    });
}

module.exports = extender.extend(pillarNewspaper, {
  initialize: initialize,
  advertisement: advertisement,
  advertisementList: advertisementList,
  newspaper: newspaper,
  newspaperList: newspaperList,
  placeAd: placeAd,
  showPaperAds: showPaperAds,
  showPapersWithAd: showPapersWithAd
});
